### PR TITLE
perf: Preserve rolling kernels after grouped row reorder

### LINKED ISSUE #28597: .rolling(...).agg(...)  ~20–40x slower ≥ 64 threads
### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import time

import polars as pl

# 250_000 rows: 1_000 groups x 250 rows. Group identity comes from the
# COMBINATION of 24 low-cardinality string keys (no single key is unique).
N_GROUPS, N_DATES, N_PAIRS, N_KEYS = 1_000, 250, 12, 24

idx = pl.int_range(0, N_GROUPS * N_DATES, eager=True)
g, d = idx // N_DATES, idx % N_DATES

data, keys, rem = {}, [], g
for i in range(N_KEYS):
    k = f"k{i}"
    keys.append(k)
    if i < N_KEYS - 1:
        card = 3 + (i % 6)
        data[k] = f"k{i}_" + (rem % card).cast(pl.String)
        rem = rem // card
    else:
        data[k] = f"k{i}_" + rem.cast(pl.String)

data["idx"] = d.cast(pl.Int64)
data["cnt"] = (((g + d) % 97) + 1).cast(pl.Int64)
for i in range(N_PAIRS):
    base = ((idx * (7919 + i * 13)) % 100003).cast(pl.Float64) / 1000.0
    data[f"m{i}"] = base - 50.0
    data[f"s{i}"] = (base % 17.0) + 0.5

df = pl.DataFrame(data)

# 24 aggregations: a count-weighted mean and stdev per column pair.
# The two families share the subexpression (cnt * m).sum() / cnt.sum().
exprs = []
for i in range(N_PAIRS):
    m, s, c = pl.col(f"m{i}"), pl.col(f"s{i}"), pl.col("cnt")
    exprs.append(((c * m).sum() / c.sum()).alias(f"wmean{i}"))
    tv = (c * (s**2 + m**2)).sum() / c.sum() - ((c * m).sum() / c.sum()) ** 2
    exprs.append(tv.sqrt().alias(f"wstd{i}"))

best = float("inf")
for _ in range(3):
    t0 = time.perf_counter()
    (
        df.lazy()
        .rolling(index_column="idx", period="753i", group_by=keys, closed="both")
        .agg(exprs)
        .collect()
    )
    best = min(best, time.perf_counter() - t0)

print(f"polars {pl.__version__}  threads={pl.thread_pool_size():3d}  best_of_3={best:.3f}s")


# for T in 32 63 64 120; do POLARS_MAX_THREADS=$T python repro.py; done
```

### Log output

```shell
polars 1.43.1  threads= 32  best_of_3=0.059s
polars 1.43.1  threads= 63  best_of_3=0.060s
polars 1.43.1  threads= 64  best_of_3=1.367s     <- 23x from one extra thread
polars 1.43.1  threads=120  best_of_3=2.230s

polars 1.34.0  threads= 63  best_of_3=0.062s
polars 1.34.0  threads= 64  best_of_3=0.062s     <- last good version: flat
polars 1.34.0  threads=120  best_of_3=0.072s
```

### Issue description

LazyFrame.rolling(index_column=..., period=..., group_by=...).agg(...) hits a hard performance cliff at exactly 64 threads: 60/61/62/63 are fast, 64/65+ are slow. The threshold does not line up with our socket boundary (60 cores/socket).

On 1.35.0 the same script at >= 64 threads panics instead:

```
thread '<unnamed>' panicked at crates/polars-mem-engine/src/executors/group_by_rolling.rs:56:9:
internal error: entered unreachable code
```

1.35.1's fix for that panic (#25075, "Panic in group_by_dynamic with group_by and multiple chunks") appears to have replaced the panic with a fallback to a much slower path, and it has been that way through 1.43.1. 

Other observations that may help locate it:

The fast path needs "enough" keys + aggregations, and both expression families. With the 24 keys fixed: 20 aggregations take ~1.3s at any thread count (no cliff — the fast path never engages), 24 aggregations take 0.06s below 64 threads and 1.3s at 64+. Either family alone (24 weighted means, or 24 weighted stdevs, or 72 plain col.quantile()s) never engages the fast path and never shows a cliff. The two families share the subexpression (cnt * m).sum() / cnt.sum(), so this smells like agg-level CSE being enabled only under some width heuristic and disabled again when the pool is >= 64.

### Expected behavior

Thread pool sizes >= 64 should perform at least as well as 63 threads on the same query (as in 1.34.0 and earlier, which were flat from 32 -> 120 threads). One extra thread should never cost 20-40x.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.43.1
Index type:          UInt32
Platform:            Linux-4.18.0-x86_64-with-glibc2.28
Python:              3.12.12 (main, Dec  9 2025, 19:02:36) [Clang 21.1.4 ]
Runtime:             rt32

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       <not installed>
boto3                <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               <not installed>
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           <not installed>
numpy                <not installed>
openpyxl             <not installed>
pandas               <not installed>
polars_cloud         <not installed>
pyarrow              <not installed>
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by raphaelroshan**
Closes #28597. (It's not accepted, do let me know if its not supposed to be a `Closes`)

Based on the analysis by @kdn36 and @0xC61 in that thread: `.rolling(..., group_by=...)` reorders rows with `DataFrame::take_unchecked`, which shards the gather by rows once `RAYON.current_num_threads() > df.width()` and leaves the columns multi-chunk. `_use_rolling_kernels` requires `chunks.len() == 1`, so from the knee (`width + 1` threads) onwards the overlapping/monotonic groups skip their stateful kernels and every window is recomputed independently — O(n·w) instead of O(n).

`mean`/`median`/`quantile` never showed the cliff because their dispatchers in `group_by/aggregations/dispatch.rs` already rechunk ("Prevent a rechunk for every individual group"); `sum`/`min`/`max`/`var`/`std` reach the gate with the raw chunk layout.

This rechunks once after the reorder, only when the groups are overlapping and monotonic and the frame actually came back chunked. Using  `rechunk_mut_par()` here since it consolidates multi chunk column into one chunk and parallelizes that across columns

## Benchmark

Apple M1 Pro, `make build-fast-release`, against `4fff6cc467`. Output checksums are
identical between branches in every run below.

#### @kdn36's MRE (knee at 5, `width=4`)

in-memory engine, median of 30 runs after 5 warmups:

| threads | 3 | 4 | 5 | 6 | 8 | 16 |
| --- | --- | --- | --- | --- | --- | --- |
| main | 10.96ms | 9.93ms | 264.15ms | 234.05ms | 187.82ms | 178.60ms |
| this PR | 11.59ms | 9.78ms | 8.57ms | 8.44ms | 9.19ms | 10.08ms |

### Original repro from the issue

Best of 3, sweeping `POLARS_MAX_THREADS`:

in-memory engine

| threads | 32 | 48 | 62 | 63 | 64 | 65 | 96 | 120 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| main | 0.077s | 0.101s | 0.088s | 0.091s | 1.488s | 1.467s | 1.483s | 1.519s |
| this PR | 0.087s | 0.086s | 0.091s | 0.079s | 0.094s | 0.087s | 0.091s | 0.114s |

streaming engine

| threads | 32 | 48 | 51 | 52 | 56 | 64 | 96 | 120 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| main | 0.087s | 0.077s | 0.121s | 1.491s | 1.503s | 1.450s | 1.666s | 1.551s |
| this PR | 0.083s | 0.101s | 0.112s | 0.089s | 0.111s | 0.105s | 0.110s | 0.106s |

The in-memory knee is at 64 as reported. The streaming engine passes a narrower frame to the same executor, so its knee is at 52. Below the knee the gather already returns a single chunk, the new rechunk never runs, and timings are unchanged on both branches.

<details>
<summary>per-aggregation benchmark (2M rows, 4 groups, in-memory)</summary>

| agg | period | main | this PR |
| --- | --- | --- | --- |
| sum | 5000i | 0.399s | 0.097s |
| sum | 20000i | 0.471s | 0.103s |
| sum | 50000i | 0.626s | 0.107s |
| min | 5000i | 0.413s | 0.110s |
| std | 5000i | 0.500s | 0.117s |
| mean | 5000i | 0.119s | 0.104s |

`mean` is the control: already rechunked, so unchanged.

</details>

## Verified locally

- `pytest tests/unit/operations/rolling tests/unit/operations/test_group_by_dynamic.py` with `POLARS_MAX_THREADS=64`: 352 passed.
- Benchmarks above were run on my own machine, not CI.

## Note for reviewers

An alternate fix would be to rechunk in ChunkedArray::agg_sum/agg_min/agg_max/agg_var (crates/polars-core/src/frame/group_by/aggregations/mod.rs), the way Series::agg_mean already does in dispatch.rs. That would also cover multi-chunk frames arriving from other executors, at the cost of rechunking per aggregation rather than once per frame. Happy to move it if you prefer that placement.

### AI disclosure
Claude was used for formatting and to run the benchmarking

### CODE REVIEW

File: py-polars/tests/unit/operations/rolling/test_rolling.py [hunk 1]
```diff
@@ -2576,3 +2576,33 @@ def test_min_periods_removed() -> None:
 
     with pytest.raises(ArgumentRemovedError, match=re.escape(msg)):
         pl.rolling_corr("a", "b", window_size=2, min_periods=1)  # type: ignore[call-arg]
+
+
+def test_rolling_group_by_interleaved_keys_28597() -> None:
```
Comment (ritchie46) [inline py-polars/tests/unit/operations/rolling/test_rolling.py:2581]: This test is not needed right? 

It was only slow, not wrong before, I would assume?

Comment (raphaelroshan) [reply py-polars/tests/unit/operations/rolling/test_rolling.py:2581]: Hi @ritchie46 , thanks for reviewing this! Yep it wasnt wrong just slow, I added the test to test this particular flow both for code coverage and to help guard against future breaks for this edge case. I can remove it if you prefer not to have it
